lib/SIL/Verifier/SILVerifier.cpp

```diff
@@ -2915,6 +2915,41 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
     require(!sd->isResilient(F.getModule().getSwiftModule(),
                              F.getResilienceExpansion()),
             "cannot access storage of resilient struct");
+    if (F.hasOwnership()) {
+      // Make sure that all of our destructure results ownership kinds are
+      // compatible with our destructure_struct's ownership kind /and/ that if
+      // our destructure ownership kind is non-trivial then all non-trivial
+      // results must have the same ownership kind as our operand.
+      auto parentKind = DSI->getOwnershipKind();
+      for (const DestructureStructResult &result : DSI->getAllResultsBuffer()) {
```

Comment:
What you're really doing here is using the parent ownership to constrain the result ownership. Now that you can convert between ValueOwnershipKind and OwnershipConstraint you can express these requirements as a single requirement:
`require(OwnershipConstraint(parentKind).satisfiedBy(result.getOwnershipKind())`
or, alternatively you can just say 
`require(parentKind.merge(result.getOwnershipKind()) == parentKind)`

Reply:
Right, I need to finish refactoring though.

I think that parentKind.isCompatibleWith is actually correct and my extra requires are unnecessary. I was worried at the time about a .none parent ownership being compatible with a result that is not .none. But that can't happen. I am going to remove that part.